### ISSUE TITLE
Screen doesn't refresh after sorting

### ISSUE THREAD

**Issue by Mikee-3000** [saulpw/visidata#1991]
**Short description**
Sort a string column

**Expected result**
View refreshes to reflect the new sort

**Actual result with screenshot**
Not all of the column is shown in the new order. "Going over" each row (i.e. with k or j) resets the shown content to the correct one.

- screen recording: https://github.com/saulpw/visidata/assets/13318998/dd572fcd-6382-472e-b710-3f4453d43175
- sample data: [sample.csv](https://github.com/saulpw/visidata/files/12303485/sample.csv)
- command log: [sample.vdj.txt](https://github.com/saulpw/visidata/files/12303502/sample.vdj.txt)

**Additional context**
Version of VisiData: saul.pw/VisiData v2.11
Version of Python: Python 3.11.4

**Comment by Mikee-3000** [comment]
I've tried different terminals (iTerm, Warp, Kitty) and different displays, also I've tried reinstalling VD using both pip and brew.

**Comment by saulpw** [comment]
Are you using tmux or other terminal intermediary?  What's your $TERM environment variable set to?

Also, for a quicker workaround, `Ctrl+L` should refresh the entire display so you don't have to scroll over each row.

**Comment by Mikee-3000** [comment]
$TERM is set to xterm-256color
I originally noticed this in a Zellij session, but it happens the same when I try in just the terminal

**Comment by saulpw** [comment]
Hm interesting.  I don't use MacOS but many people do, and I imagine we would have heard about it before now if this was the case.  (Can other people chime in with their terminals/versions?)  FWIW this doesn't happen for me on Linux in any bare terminal emulator I've tried (kitty, xterm, urxvt).  It does happen inside tmate, though tmate seems to have refresh issues in many apps (particularly with tmux split panes) and so I've gotten used to frequent `Ctrl+L`.

Also please use the latest develop version of VisiData and see if that makes any difference.

**Comment by Mikee-3000** [comment]
I've tried with v2.12dev version of VisiData, It's still the same problem. 
But I agree, if it's not reported by anyone else, then it's probably something in my setup, not related to VD.
Thanks for the Ctrl+L shortcut, that's much easier than running through all the rows.

**Comment by midichef** [comment]
@Mikee-3000 as a temporary workaround, you can try putting this code in your .visidatarc file. It will redraw the screen after sorting finishes.

It's just the same code as in Visidata's `Sheet.sort()`, with two lines added to redraw the screen. I can't test it under your conditions, but it seems to work on my system and not cause any problems.

```
# modified version of sort() to redraw the screen after sort finishes
# temporary workaround for https://github.com/saulpw/visidata/issues/1991
@Sheet.api
@asyncthread
def sort(self):
    'Sort rows according to the current internal ordering.'
    if self.rows is UNLOADED:
        return
    try:
        with Progress(gerund='sorting', total=self.nRows) as prog:
            # must not reassign self.rows: use .sort() instead of sorted()
            self.rows.sort(key=lambda r,self=self,prog=prog: self.sortkey(r, prog=prog))
    except TypeError as e:
        vd.warning('sort incomplete due to TypeError; change column type')
        vd.exceptionCaught(e, status=False)
    # The above definition of sort() is copied from sort.py (line 54).
    # Now we add the actions that happen for Ctrl+L.
    # Taken from the definition of the 'redraw' command in sheets.py (line 1085):
    vd.redraw();
    self.refresh()
```

**Comment by Mikee-3000** [comment]
This works, thank you so much!Sent from my iPhoneOn 11 Aug 2023, at 22:07, midichef ***@***.***> wrote:﻿
@Mikee-3000 as a temporary workaround, you can try putting this code in your .visidatarc file. It will redraw the screen after sorting finishes.
It's just the same code as in Visidata's Sheet.sort(), with two lines added to redraw the screen. I can't test it under your conditions, but it seems to work on my system and not cause any problems.
# modified version of sort() to redraw the screen after sort finishes
# temporary workaround for https://github.com/saulpw/visidata/issues/1991
@Sheet.api
@asyncthread
def sort(self):
    'Sort rows according to the current internal ordering.'
    if self.rows is UNLOADED:
        return
    try:
        with Progress(gerund='sorting', total=self.nRows) as prog:
            # must not reassign self.rows: use .sort() instead of sorted()
            self.rows.sort(key=lambda r,self=self,prog=prog: self.sortkey(r, prog=prog))
    except TypeError as e:
        vd.warning('sort incomplete due to TypeError; change column type')
        vd.exceptionCaught(e, status=False)
    # The above definition of sort() is copied from sort.py (line 54).
    # Now we add the actions that happen for Ctrl+L.
    # Taken from the definition of the 'redraw' command in sheets.py (line 1085):
    vd.redraw();
    self.refresh()

**Comment by saulpw** [comment]
Okay so I **can** repro this in my local setup with this dataset at least by running inside of tmux with kitty.  So I spent a couple hours today trying to fix this without getting a lot of flicker (which would be a regression).  Things I've learned:

- It's not good to do a `vd.redraw()` or indeed any curses action inside another thread; this will result in garbage being sent to the screen as multiple threads get their escape sequences commingled to stdout.
- maybe `BaseSheet.refresh()` itself should have null semantics, instead of updating the screen, existing only to allow sheet subclasses like Canvas to override the implementation and recalculate internal state for their own draw methods.
- The current clear/refresh in `BaseSheet.refresh()` may not be necessary anyway; the functionality in `vd.redraw()` should cause all windows on the screen to completely redraw.
- Notably, removing `@asyncthread` from `Sheet.sort()` does not change this behavior, so it doesn't have anything to do with async as I was expecting.

So, there must be some way to get the screen to update properly without clearing it entirely.  The current code seems to work well in this regard in almost all cases; I haven't found any commands besides sort which has this issue.  I don't know why sort is special; it'd be great if we could find anything else that reliably exhibits this behavior. I would guess it has something to do with subwindows, and the solution will involve some curses operation like `scr.refresh()`, `scr.redrawwin()`, `scr.syncup()`, `scr.touchwin()`, or `curses.doupdate()`.  `scr.clear()` or `scr.erase()` are heavy-handed and will get the job done, at the cost of noticeable flicker, which is fine for a manual Ctrl+L, but not for anything that's called after a command or every draw cycle.
